psycopg/psycopgmodule.c

```diff
@@ -112,6 +112,44 @@ psyco_connect(PyObject *self, PyObject *args, PyObject *keywds)
     return conn;
 }
 
+#define psyco_parse_dsn_doc \
+"parse_dsn(dsn) -- Parse database connection string.\n\n"
+
+static PyObject *
+psyco_parse_dsn(PyObject *self, PyObject *args)
+{
+    char *dsn, *err;
+    PQconninfoOption *options = NULL, *o;
+    PyObject *res = NULL, *value;
+
+    if (!PyArg_ParseTuple(args, "s", &dsn)) {
+        return NULL;
+    }
+
+    options = PQconninfoParse(dsn, &err);
+    if (!options) {
+        PyErr_Format(PyExc_RuntimeError, "PQconninfoParse: %s: %s", dsn, err);
```

Comment:
This error is wrong. The dsn shouldn't be printed back as it may contain passwords that may end up in the logs. It shoud be a ProgrammingError if the dsn is wrong (with a human error message such as "error parsing the dsn: %s" % err) or an OperationalError if err is null as that's a memory allocation error according to the docs).